mangaki/mangaki/algo/als2.py

```diff
@@ -0,0 +1,111 @@
+from collections import defaultdict
+
+import numpy as np
+
+from mangaki.algo.recommendation_algorithm import RecommendationAlgorithm, register_algorithm
+
+
+@register_algorithm('als2', {'nb_components': 20})
+class MangakiALS2(RecommendationAlgorithm):
+    M = None
+    U = None
+    VT = None
+    def __init__(self, nb_components=20, nb_iterations=20, lambda_=0.1):
+        super().__init__()
+        self.nb_components = nb_components
+        self.nb_iterations = nb_iterations
+        self.lambda_ = lambda_
+
+    @property
+    def is_serializable(self):
+        return True
+
+    def load(self, filename):
+        backup = super().load(filename)
+        self.M = backup.M
+        self.U = backup.U
+        self.VT = backup.VT
+        self.means = backup.means
+
+    def make_matrix(self, X, y):
+        matrix = defaultdict(dict)
+        means = np.zeros((self.nb_users,))
+        for (user, work), rating in zip(X, y):
+            matrix[user][work] = rating
+            means[user] += rating
+        for user in matrix:
+            means[user] /= len(matrix[user])
+        return matrix, means
+
+    def fit_user(self, user, matrix):
+        Ru = np.array(list(matrix[user].values()))
+        Ju = list(matrix[user].keys())
+        Wu = self.W_item[Ju]
+        wu = self.w + self.W_user[user]
+        Vu = self.VT[:, Ju]
+        Gu = self.lambda_ * len(matrix[user]) * np.eye(self.nb_components)
+        self.U[[user],:] = np.linalg.solve(Vu.dot(Vu.T) + Gu, Vu.dot(Ru - Wu - wu)).T
+        self.W_user[user] = (Ru - Vu.T.dot(self.U[user, :]) - Wu).mean() / (1 + self.lambda_) - self.w
+
+    def fit_work(self, work, matrixT):
+        Ri = np.array(list(matrixT[work].values()))
+        Ii = list(matrixT[work].keys())
+        Wi = self.W_user[Ii]
+        wi = self.w + self.W_item[work]
+        Ui = self.U[Ii, :].T
+        Gi = self.lambda_ * len(matrixT[work]) * np.eye(self.nb_components)
+        self.VT[:,[work]] = np.linalg.solve(Ui.dot(Ui.T) + Gi, Ui.dot(Ri - Wi - wi).reshape(-1, 1))
+        self.W_item[work] = (Ri - Ui.T.dot(self.VT[:, work]) - Wi).mean() / (1 + self.lambda_) - self.w
+
+    def factorize(self, matrix, random_state):
+        # Preprocessings
+        matrixT = defaultdict(dict)
+        for user in matrix:
+            for work in matrix[user]:
+                matrixT[work][user] = matrix[user][work]
+        # Init
+        self.U = np.random.rand(self.nb_users, self.nb_components)
+        self.VT = np.random.rand(self.nb_components, self.nb_works)
+        self.w = 0.#np.random.random()
```

Comment:
Please remove the comment, it's confusing. Either, add a flag to do random initialization for `self.w`, either set it to zero definitely.

Reply:
Udon.